tests/unit/machine/all_machines_unittest.cc

```diff
@@ -0,0 +1,386 @@
+		{
+			memcpy(X_subsampled.get_column_vector(i), X.get_column_vector(subset[i]), X.num_rows * sizeof(decltype(X(0,0))));
+			y_subsampled[i] = y[subset[i]];
+		}
+
+		auto features_subsampled = new CDenseFeatures<float64_t>(X_subsampled);
+		auto labels_subsampled = make_clone(data.second);
+		labels_subsampled->set_labels(y_subsampled);
+
+		auto machine_subset = make_clone(machine);
+		auto machine_subsampled = make_clone(machine);
+		if (machine->has("seed"))
+		{
+			machine_subset->put("seed", 1);
+			machine_subsampled->put("seed", 1);
+		}
+
+		machine_subset->set_labels(labels_subset);
+		machine_subset->train(features_subset);
+		machine_subsampled->set_labels(labels_subsampled);
+		machine_subsampled->train(features_subsampled);
+
+		auto result_subset = machine_subset->apply(data.first);
+		auto result_subsampled = machine_subsampled->apply(data.first);
+
+		EXPECT_TRUE(result_subset->equals(result_subsampled));
+	}
+}
+
+TEST(AllMachines, cv_thread_consistency)
+{
+	std::set<std::string> ignores = {
+			"RandomForest" // BaggingMachine.cpp line 153: Number of bag is not set!"
+			};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto data = generate_data(machine);
+		auto cv = generate_cv(machine, data);
+		auto cv2 = make_clone(cv);
+
+		get_global_parallel()->set_num_threads(1);
+		cv->put("seed", 1);
+		auto result_single = cv->evaluate();
+
+		get_global_parallel()->set_num_threads(4);
+		cv2->put("seed", 1);
+		auto result_multi = cv2->evaluate();
+
+		EXPECT_TRUE(result_single->equals(result_multi));
+	}
+}
+
+TEST(AllMachines, train_apply_no_side_effects)
+{
+	std::set<std::string> ignores = {
+			"RandomForest" // BaggingMachine.cpp line 153: Number of bag is not set!"
+			};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto data = generate_data(machine);
+
+		auto features_before = data.first->clone();
+		auto labels_before = data.second->clone();
+
+		machine->set_labels(data.second);
+		machine->train(data.first);
+		machine->apply(data.first);
+
+		auto features_after = data.first->clone();
+		auto labels_after = data.second->clone();
+
+		EXPECT_TRUE(features_before->equals(features_after));
+		EXPECT_TRUE(labels_before->equals(labels_after));
+	}
+}
+
+TEST(AllMachines, trained_model_serialization_consistency)
+{
+	std::set<std::string> ignores = {
+			"RandomForest" // BaggingMachine.cpp line 153: Number of bag is not set!"
+			};
+	for (auto obj : sg_object_iterator<untemplated_sgobject>(all_machines).ignore(ignores))
+	{
+		auto machine = obj->as<CMachine>();
+		SCOPED_TRACE(machine->get_name());
+
+		init_machine(machine);
+		auto data = generate_data(machine);
+
+		machine->set_labels(data.second);
```

Comment:
the triplet: set_labels, train, apply sprayed around this source again... while it could be a function that returns a tuple that you nicely tie and get it back

Reply:
for further info: https://en.cppreference.com/w/cpp/language/structured_binding

Reply:
but in this case you would actually just return the predictions... so you dont even need that... just pass the rvalue of the tuple (labels, data)